Improvements for "create collage" feature

discharged-spider opened the pull request:
Following: https://github.com/T8RIN/ImageToolbox/issues/2154

Also, I've realized that this PR is quite **huge**, so here’s the main logic:

1. I've introduced **Handles** to allow dragging borders between collage windows.
2. But one wall may influence multiple windows, so actually we have:
   values <- collage params (a.k.a. degrees of freedom)  ~ positions of all walls that can move.
   These values are managed by **ParamsManager**, which knows how to update the necessary windows when **some values are changed** (so all updates are local, and we don’t recompute the whole layout when dragging walls).
3. Each window provides **Handles**. Each handle may change exactly one value managed by this handle - usually one of the walls touched by this window.
4. But what happens if the user drags a handle into an illegal state - for example, making a window zero-sized? When applying updates, the Manager can catch an `InvalidValue` exception and roll back all updates (this way, each handle is restricted from moving to invalid positions).
5. Finally, I’ve created the `buildParamsCollage` helper function that makes it easier to register all necessary params (param -> value from p.2), and describe how each window should be computed based only on the necessary values. Each window should also provide its handles, but for **Boxed** windows all handles may be calculated automatically (see `addBoxedItem`).

The introduced mechanism is really flexible and can be used for all types of dynamic templates (for example, some custom handle may even change the radius of a circular window). For now, however, I’ve decided to use it only for simple boxed layouts.

Also, there is some fairly complex `updateFrame` logic, but that’s because we want the images to stay almost centered in the frame when adjusting window sizes. And we definitely don’t want to introduce extra background space if it wasn’t there before.

discharged-spider commented:
I checked the current app (without my modifications), screen rotation is disabled for "create collage", it is always in portrait mode.

discharged-spider commented on libs/collages/src/main/java/com/t8rin/collages/Collage.kt:
```diff
@@ -128,9 +149,14 @@ fun Collage(
                     }
                 },
                 update = {
-                    if (previousSize != size || it.mPhotoItems != imagesMapped || needToInvalidate || previousAspect != aspectRatio || previousScale != outputScaleRatio) {
```
With this section included, I have a bug where sometimes changing the "disableRotation" flag triggers an update on Collage, and then all image positions and rotations are reset.

I may be wrong, but my intuition is that this is happening because PhotoItem can now change dynamically (its bounds will be updated during dragging), and we don’t want to compare it directly with imagesMapped.

For changes that include image updates, we have the needToInvalidate flag, which is set inside LaunchedEffect, so Collage will be updated correctly when it's necessary.

T8RIN commented:
Okay, i think it's better if you try again, and if you will not be able to do that i will merge and continue myself👀

discharged-spider commented:
I had a bit of free time today, so I fixed the layout preservation problem.
I actually found out that in the current version all image positions are reset when:

* the layout changes (because the component size in px is probably different),
* the output size is changed — this one is the strangest, and probably disappointing for users when they just try to change the export resolution and all progress is reset,
* the collage’s aspect ratio is changed.

So, I rewrote the **Collage** component update logic. Now it readapts the existing collage to new changes (fixing all the points above).
On top of that, I found a lot of code duplication related to `outputScale` (for example, `mScaleMatrix` always equals `ImageMatrix.postScale(scale, scale)` but was being recomputed in many places), so I refactored it.

I tested the current version — the layout is preserved, and I’m happy with how the collage size adaptation works.
